Validator/src/main/java/org/openzen/zenscript/validator/visitors/ReturnStatementValidator.java

```diff
@@ -0,0 +1,184 @@
+
+		return ReturnKind.branches(onThen, onElse);
+	}
+
+	@Override
+	public ReturnKind visitLock(LockStatement statement) {
+		return statement.content.accept(this);
+	}
+
+	@Override
+	public ReturnKind visitReturn(ReturnStatement statement) {
+		return ReturnKind.ALWAYS_RETURNS;
+	}
+
+	@Override
+	public ReturnKind visitSwitch(SwitchStatement statement) {
+		return statement.cases.stream()
+				.map(c -> mergeInsideBlock(c.statements))
+				.reduce(ReturnKind::branches)
+				.orElse(ReturnKind.NEVER_RETURNS);
+	}
+
+	@Override
+	public ReturnKind visitThrow(ThrowStatement statement) {
+		return ReturnKind.NEVER_RETURNS;
+	}
+
+	@Override
+	public ReturnKind visitTryCatch(TryCatchStatement statement) {
+		ReturnKind content = statement.content.accept(this);
+		ReturnKind catchClauses = statement.catchClauses.stream()
+				.map(it -> it.content.accept(this))
+				.reduce(ReturnKind::branches)
+				.orElse(ReturnKind.NEVER_RETURNS);
+		ReturnKind finallyClause = statement.finallyClause == null ? ReturnKind.NEVER_RETURNS : statement.finallyClause.accept(this);
+
+		// If finally clause always returns it wins, otherwise use individual content/catch clauses
+		return ReturnKind.max(finallyClause, ReturnKind.branches(content, catchClauses));
+	}
+
+	@Override
+	public ReturnKind visitVar(VarStatement statement) {
+		return ReturnKind.NEVER_RETURNS;
+	}
+
+	@Override
+	public ReturnKind visitWhile(WhileStatement statement) {
+		return statement.content.accept(this);
+	}
+
+	private ReturnKind mergeInsideBlock(Statement[] statements) {
+		ReturnKind result = ReturnKind.NEVER_RETURNS;
+		for (Statement statement : statements) {
+			if (result == ReturnKind.ALWAYS_RETURNS) {
+				validator.logError(statement.position, CompileErrors.unreachableStatement());
+				break;
+			}
+
+			result = ReturnKind.max(result, statement.accept(this));
+		}
+
+		return result;
+	}
+
+	public enum ReturnKind {
+		/**
+		 * No return statements are present
+		 */
+		NEVER_RETURNS,
+		/**
+		 * Only some cases return (e.g. in the if-branch but not inside the else branch, or inside a try but not in the catch)
+		 */
+		RETURNS_IN_SOME_CASES,
+		/**
+		 * All possible branches return a value
+		 */
+		ALWAYS_RETURNS;
+
+		/**
+		 * Used to merge returnKinds of sequential statements into one (e.g. all statements inside a block)
+		 * In that case, the highest kind wins (ignoring duplicate return statements for now).
+		 */
+		public static ReturnKind max(ReturnKind a, ReturnKind b) {
+			return a.ordinal() > b.ordinal() ? a : b;
+		}
+
+		/**
+		 * Used to merge returnKinds of different branches, e.g. inside an if/else branch or switch/cases.
+		 */
+		public static ReturnKind branches(ReturnKind a, ReturnKind b) {
+			if (a == NEVER_RETURNS && b == NEVER_RETURNS) {
+				return NEVER_RETURNS;
+			}
+			if (a == ALWAYS_RETURNS && b == ALWAYS_RETURNS) {
+				return ALWAYS_RETURNS;
+			}
+			return RETURNS_IN_SOME_CASES;
+		}
```

Comment:
💭 Discussion: The same functionality would be achieved by the code block below.  
I find the explicit if/else structure more readable, what do you think?

```suggestion
		public static ReturnKind branches(ReturnKind a, ReturnKind b) {
			return a == b ? a : RETURNS_IN_SOME_CASES;
		}
```